pkg/adaptor/proxy/proxy.go

```diff
@@ -117,38 +113,28 @@ func (p *agentProxy) dial(ctx context.Context, address string) (net.Conn, error)
 		dialer = &net.Dialer{}
 	}
 
-	maxRetries := defaultMaxRetries
-	count := 1
-	for {
-		var err error
-
-		func() {
-			ctx, cancel := context.WithTimeout(ctx, p.retryInterval)
-			defer cancel()
+	ctx, cancel := context.WithTimeout(ctx, time.Duration(p.proxyTimeout)*time.Minute)
+	defer cancel()
 
+	logger.Printf("Trying to establish agent proxy connection to %s", address)
+	err := retry.Do(
+		func() error {
+			var err error
 			conn, err = dialer.DialContext(ctx, "tcp", address)
-
-			if err == nil || count == maxRetries {
-				return
+			if err != nil {
+				return err
 			}
-			<-ctx.Done()
-		}()
+			return nil
+		},
+		retry.Context(ctx),
+	)
 
-		if err == nil {
-			break
-		}
-		if count == maxRetries {
-			err := fmt.Errorf("reaches max retry count. gave up establishing agent proxy connection to %s: %w", address, err)
-			logger.Print(err)
-			return nil, err
-		}
-		logger.Printf("failed to establish agent proxy connection to %s: %v. (retrying... %d/%d)", address, err, count, p.maxRetries)
-
-		count++
+	if err != nil {
+		logger.Printf("failed to establish agent proxy connection to %s", address)
+		return nil, err
```

Comment:
Could you wrap the original error like this?
```
err = fmt.Errorf("failed to establish agent proxy connection to %s: %w", address, err)
logger.Print(err)
return nil, err
```

If the error message becomes too long, you can specify the `LastErrorOnly` option.
https://pkg.go.dev/github.com/avast/retry-go#LastErrorOnly